### PR TITLE
Support for order conversion on Arrays

### PR DESCRIPTION
**Pull request by Thrameos**
@pelson  If the goal is simply to change the reported type to "l" then wouldn't this be simpler.

The problem of course is that direct buffers are still type "q".  But if you try to fix that you get this error...

"RuntimeError: Item size 8 for PEP 3118 buffer format string >l does not match the dtype i item size 4."

Because long is supposed to be 4 not 8.   So the correct type should be "q" for Java long.

### CODE REVIEW

**Comment by Thrameos** [comment]
@pelson This implements the order conversions and fixes the issue with long.  Unfortunately it exposes other issues.   On windows "i" type buffers go to "intc" rather than "int32".   I can make it into "=i" which will go to "int32", but then memoryview writes will fail.   Same goes for long.   Thus we would have to remove the memoryview write functionality.

**Comment by Thrameos** [comment]
@pelson I added test casts for every combination that I am able to hit.   Some are possible in the specification (network order, pascal byte string, etc), but I have no clue how to create a buffer with such monstrosities. 

We unfortunately broke one behavior which is direct writing to native buffers because once you place the standard "=i" versus native "@i" you can't cast the memory buffer.  This is a rare case, but it was possible before. 

I hope this resolves the issue.

**Comment by marscher** [comment]
> We unfortunately broke one behavior which is direct writing to native buffers because once you place the standard "=i" versus native "@i" you can't cast the memory buffer. This is a rare case, but it was possible before.

But this rare case is not being used to create writable Numpy arrays upon those memoryviews, right? Since we provide our own interface or use a different pathway.

Could somebody make the effort to test this with their specific use cases, e.g. the mentioned "exotic" memory layouts?

**Comment by Thrameos** [comment]
The only behavior that we broke was the very unusual one in which you create a memoryview from an existing array and then proceed to modify the contents of that view which alters the contents of the original Java array.   Basically this is to create shared memory.

The paths to move data back and forth are using memory view as well, but as you are only copying there are no limitations.   Unfortunately this is one of those we can only have things one way or the other.  If we properly state the memory which is needed such the numpy arrays will correctly get the size then we must force the type with "=i".  If we do so then it isn't a native primitive and thus the Python memoryview object won't let you alter the bytes.  You can force the issue by applying a series of casts, but this would require one to alter the code.   Those casts would be backwards compatible, it is just we didn't require them before.    And this restriction is really coming from Python not us.   There is no reason that they can't implement byte level diddling of memoryviews which numpy supports. 

See line 77 of test/jpypetest/test_buffer.py for the required casting.
```
        b = memoryview(ja)  # Create a view  (BEFORE)
        b = memoryview(ja).cast("B")  # Create a view (AFTER, must cast view to allow modifications)
        b[0] = 123
```

I find it very sad that we are still playing out bad decisions from C compilers 30 years ago, when they decided that long == int but with a different name, which is what is driving this as Python ends up with "i" int  (4), "l" long (4 or 8), "q" (8 or longer).   Bad decisions never go away... they just fester.

Of course most of this is pretty exotic.   I am guessing that unless CellProfiler or Bioformats are making use of memoryview direct access, that it was a feature that I added just for completeness.  I wouldn't have noticed it if I hadn't placed a test of almost all the memoryview behaviors, and I doubt it is anywhere in the documentation.

The main change here is that we now report our buffers with fixed sizes and never with native types.  This makes numpy happy as there is no ambiguity about the type.  Of course I really don't know what the difference between type="l" size=8 and type="q" size=8, but numpy thinks that only one should go to np.dtype.int64.  I also for completeness added support for converting from all endianness.  This means that any numpy array (with the exception of a "short" float type "e") should now run through our converter code successfully.   I added tests for all the endian changes that were supported by native python.  There are a few others that may be possible either through numpy or some other tool, but I could figure out how to produce the required inputs to test.   If I could I would test it... just because I am the ultimate completionist.

**Comment by Thrameos** [comment]
Oh one last thing.   This is really architecture dependent.   On windows numpy would take the specification to one dtype, and linux would do something different.   Thus it was very hard to actually make a PR which got every type to be consistent across architectures.  It is very possible that some other platform we are not testing goes to something different.   But this isn't our issue, but rather  numpy differences.

**Comment by marscher** [comment]
Thanks for you detailed explainations, I think I get the picture now. I've searched the code of CellProfiler and Bioformats for usages of memoryview, and there is no access pattern you described in there.

So I conclude that this PR is filling a gap. The other PR #1039 is handling this in a different manner, which you @Thrameos do not think is the right solution. I actually did not find the time to understand the other solution in detail, since it changed a lot of code. Actually I'd prefer to keep things simple and your provided solution keeps it that way.

Eventually we should never have fiddle around with our provided jni.h, which actually derefers the defintion of integral types (jbyte, jint, jlong) to jni_md.h. Somehow we have included (maschine-dependent) types in our main jni.h file. I guess this was a mistake, since these definitions differ depending on the used architecture.
The (bad?) deciscion to provide jni.h (with the included integral type defs) was made to simply jpypes setup. At least on conda-forge there is JDK provided, which also provides the platform dependent jni_md.h, so we wouldn't need that.
Providing binaries on PyPI should work similar, if we stick to holy build box and OpenJDK.

But for more exotic configurations (HP-Ux etc.) we should definetly prefer the external JDK jni headers.

**Comment by Thrameos** [comment]
Sorry about that i meant to edit the body of the first comment to inckude a summary.  In the middle of a delivery so time pressed, unfortunately alsio means I need to get 3.10 version jpype out door.  However I wont do that until all segfaults are cleared and this request and the other "glorious segfault" are holding that up 

What is addressed here.
- adds support for all buffer transfer requests including different byte orders.
-  changes memory view types to fixed sizes to ensure numpy wrappers are int32 and int64 on all tested platforms.
- test cases for all buffer types.
- fixed segfault when requesting exotic buffer transfer requests.  (One test segfaulted old code)

The other PR changed a lot of code and addressed 2 of these issues.  But the test bench was broken at the time, so I had having a hard time following as well as the PR title was for the second but the code changes were for the 4th.  My main reluctance is that seveeal tests were broken and it seemed like a partial solution so i decided to try for something dealing with all.

**Review by marscher — APPROVED** [review]

**Comment by pelson** [comment]
This MR is evolved a lot from the original commit it had nicely. It is basically the same change as https://github.com/jpype-project/jpype/pull/1039, except this one handles byte-swapping, but doesn't handle the float/double types (which shouldn't be platform dependent either - though this is not really something that is a real problem from what I can see).

There are some important comments in #1039 that we should take note of, which related to discussions in this MR:

> Eventually we should never have fiddle around with our provided jni.h

https://github.com/jpype-project/jpype/pull/1039#discussion_r830143672

> We unfortunately broke one behavior which is direct writing to native buffers because once you place the standard "=i" versus native "@i" you can't cast the memory buffer. This is a rare case, but it was possible before.

https://github.com/jpype-project/jpype/pull/1039#discussion_r830953513

----

There are also a bug fix which we should try to keep which didn't make it into this MR:

 * There is a wrong type definition in jp_inttype: https://github.com/jpype-project/jpype/pull/1039#discussion_r830949403


In summary:

* This MR is great. Thanks for doing it @Thrameos. :+1:

**Comment by Thrameos** [comment]
Would you mind submitting a PR with additional tests in test/jpypetest for the behaviors that you see as missing? You just need to put tests that are failing, as I can use that to fix it when I cycle back to glorious segfault.   I thought that I got the int type and the long type changes, but if I overlooked a behavior I can revise.
